mscore/propertymenu.cpp

```diff
@@ -588,5 +558,84 @@ void ScoreView::editTremoloBarProperties(TremoloBar* tb)
+                              // If instrument change is at the start of a measure, use the measure as the element, as this will place the instrument change before the barline.
+                              Element* element = ic->rtick().isZero() ? toElement(ic->findMeasure()) : toElement(ic);
+                              score()->undoChangeClef(part->staff(i), element, clefType, ic);
+                              }
+                        }
+                  // Change key signature if necessary
+                  if (it->transpose != oldV) {
+                        for (int i = 0; i < part->nstaves(); i++) {
+                              if (!part->staff(i)->keySigEvent(tickStart).isAtonal()) {
+                                    KeySigEvent ks;
+                                    ks.setForInstrumentChange(true);
+                                    Key key = part->staff(i)->key(tickStart);
+                                    if (!score()->styleB(Sid::concertPitch))
+                                          key = transposeKey(key, oldV);
+                                    ks.setKey(key);
+                                    score()->undoChangeKeySig(part->staff(i), tickStart, ks, ic);
+                                    }
+                              }
+                        }
+                  // change instrument in all linked scores
+                  for (ScoreElement* se : ic->linkList()) {
+                        InstrumentChange* lic = static_cast<InstrumentChange*>(se);
+                        Instrument* instrument = new Instrument(Instrument::fromTemplate(it));
+                        lic->score()->undo(new ChangeInstrument(lic, instrument));
+                        }
+                  
+                  // transpose for current score only
+                  // this automatically propagates to linked scores
+                  if (part->instrument(tickStart)->transpose() != oldV) {
+                        auto i = part->instruments()->upper_bound(tickStart.ticks());    // find(), ++i
+                        Fraction tickEnd;
+                        if (i == part->instruments()->end())
+                              tickEnd = Fraction(-1, 1);
+                        else
+                              tickEnd = Fraction::fromTicks(i->first);
+                        ic->score()->transpositionChanged(part, oldV, tickStart, tickEnd);
+                        }
+			
+                  if (ic->warning()) {
+                        ic->warning()->setPlainText(it->trackName);
+                        }
+                  else if (InstrumentChange* prevIc = score()->prevInstrumentChange(ic->segment()->prev1(), ic->staff(), true)) {
+                        StaffText* staffText = prevIc->warning();
+                        prevIc->setWarning(nullptr);
+                        staffText->setPlainText(it->trackName);
+                        ic->setWarning(staffText);
```

Comment:
Is all of this logic correct? To me, it looks like this happens:

- The warning text of the instrument change before `ic`, `prevIc`, is moved to `ic`.

Why is this desirable? If `ic` doesn't have a warning text, don't you want to create a new one for it? Also, doesn't this code also assume that `prevIc` has warning text? What if it doesn't?

If you could explain your thinking here, that'd be great :)

Reply:
This is to handle a very unlikely edge case I came across in testing, to handle the user putting an instrument change between an existing instrument change and it's warning (and therefore the first note of the passage). So this means that it assigns the existing warning for the previous instrument change to the new one, and updates the text to match the new one. Like I said, a fairly uncommon use case, but it should still have the correct behaviour

Reply:
I've also just noticed that I should check if the previous instrument change even has a warning, as if the user adds two instrument changes in succession, with no notes afterwards, I think it might cause a null pointer exception

Reply:
> I've also just noticed that I should check if the previous instrument change even has a warning

Yep, as I pointed out above ;)

Also, I still don't think the logic's right. As far as I can tell, it only checks if there is a previous instrument change, but not if that instrument change's warning text comes after `ic`. So, with the current logic, this branch will always run if there's an instrument change beforehand, no matter if `ic` is before or after it's warning text. Is that right?

Reply:
The final parameter of the prevInstrumentChange means that when it looks backwards it returns nullptr if there are any notes before reaching an instrument change. If there is a note, assuming the rest of the logic is correct, that note will have the warning attached to it, and will be guaranteed to be fore the added instrument change. I should probably add a check in anyway, just in case the user has moved it or something has broken, but I'm fairly sure the base logic is correct.

Reply:
Ok, I think that makes sense...